### PR TITLE
WIP compatibility with `firebase@exp`

### PR DESCRIPTION
**Pull request by jamesdaniels**
Work on `rxfire@exp` which is compatible with `firebase@exp` is ongoing on the [`exp` branch](https://github.com/FirebaseExtended/rxfire/tree/exp).

### CODE REVIEW

**Comment by codechips** [comment]
Hi! Will this PR be merged and published to NPM under beta? I would like to try and convert my app to new Firebase SDK, but RxFire is currently blocking me from it :smiley_cat:

**Comment by jamesdaniels** [comment]
> Hi! Will this PR be merged and published to NPM under beta? I would like to try and convert my app to new Firebase SDK, but RxFire is currently blocking me from it 😺

No commitment on timing, but we're getting close. Feel free to pin to the canary artifact that's been published to NPM for now, latest is `0.500.0-canary.0df3466`

File: firestore/fromRef.ts [hunk 1]
```diff
@@ -15,44 +15,21 @@
...
+import { onSnapshot } from 'firebase/firestore';
+import { Observable } from 'rxjs';
+import { DocumentReference, DocumentData, SnapshotListenOptions, Query, DocumentSnapshot, QuerySnapshot } from './interfaces';
 
 /* eslint-disable @typescript-eslint/no-explicit-any */
-function _fromRef(
-    ref: any,
-    options: SnapshotListenOptions | undefined,
+export function fromRef<T=DocumentData>(ref: DocumentReference<T>, options?: SnapshotListenOptions): Observable<DocumentSnapshot<T>>;
+export function fromRef<T=DocumentData>(red: Query<T>, options?: SnapshotListenOptions): Observable<QuerySnapshot<T>>;
```
Comment (jhuleatt) [inline firestore/fromRef.ts:25]: caught a small typo:

```suggestion
export function fromRef<T=DocumentData>(ref: Query<T>, options?: SnapshotListenOptions): Observable<QuerySnapshot<T>>;
```

File: docs/database.md [hunk 2]
```diff
@@ -8,20 +8,20 @@ The `object()` function creates an observable that emits object changes.
 |                 |                                          |
 |-----------------|------------------------------------------|
 | **function**    | `object()`                               |
-| **params**      | `database.Reference`                     |
+| **params**      | `import('firebase/database').Reference`  |
 | **import path** | `rxfire/database`                        |
```
Comment (jhuleatt) [inline docs/database.md:12]: I get the following error when trying to import from `rxfire/database`:

```
Cannot find module 'rxfire/database' or its corresponding type declarations.
```

Comment (jhuleatt) [reply docs/database.md:12]: nevermind, I was pulling in an older build.

**Comment by uny** [comment]
I am trying this `6.0.0-canary.da26625` in Next.js, but getting an error.
```
/workspace/node_modules/rxfire/auth/index.esm.js:1
import { onAuthStateChanged, onIdTokenChanged, getIdToken } from 'firebase/auth';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:984:16)
    at Module._compile (internal/modules/cjs/loader.js:1032:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
    at Module.load (internal/modules/cjs/loader.js:933:32)
    at Function.Module._load (internal/modules/cjs/loader.js:774:14)
    at Module.require (internal/modules/cjs/loader.js:957:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.rxfire/auth (/workspace/rxfire-nextjs/.next/server/pages/index.js:66:18)
    at __webpack_require__ (/workspace/rxfire-nextjs/.next/server/webpack-runtime.js:33:42)
    at eval (webpack-internal:///./pages/index.js:11:69)
    at Object../pages/index.js (/workspace/rxfire-nextjs/.next/server/pages/index.js:22:1)
    at __webpack_require__ (/workspace/rxfire-nextjs/.next/server/webpack-runtime.js:33:42)
    at __webpack_exec__ (/workspace/rxfire-nextjs/.next/server/pages/index.js:76:52)
    at /workspace/rxfire-nextjs/.next/server/pages/index.js:77:28
    at Object.<anonymous> (/workspace/rxfire-nextjs/.next/server/pages/index.js:80:3)
    at Module._compile (internal/modules/cjs/loader.js:1068:30)
```

## Steps to reproduce
1. npx create-next-app
2. npm install rxjs rxfire@canary firebase@exp
3. Replace pages/index.js into this:
    ```javascript
    import { getApp, getApps, initializeApp } from "firebase/app";
    import { getAuth } from "firebase/auth";
    import { authState } from "rxfire/auth";

    const firebase = (() => {
      if (getApps().length === 0) {
        return initializeApp({ /* OPTIONS */ });
      } else {
        return getApp();
      }
    })();

    export default function Home() {
      React.useEffect(() => {
        const subscriptions = [];
        subscriptions[subscriptions.length] = authState(getAuth(firebase)).subscript(console.log);
        return () =>
          subscriptions.forEach((subscription) => subscription.unsubscribe());
      }, []);
      return (
        <></>
      )
    }
    ```

Do you have any idea about it?

**Comment by uny** [comment]
1. `git submodule add https://github.com/FirebaseExtended/rxfire.git rxfire` -> https://github.com/FirebaseExtended/rxfire/tree/da26625dd3448a265262af0d0537c191e20329d0
2. `(cd rxfire && npm install --force && npm run build && npm link)`
3. `npm link rxfire`

With these steps, it works well without any errors.

**Comment by uny** [comment]
I found a problem in `rxfire/firestore`.

```
Uncaught FirebaseError: Expected type '_a', but it was: an object
    at ua (webpack-internal:///../rxfire/node_modules/@firebase/firestore/dist/exp/index.browser.esm2017.js:14672:19)
    at hh (webpack-internal:///../rxfire/node_modules/@firebase/firestore/dist/exp/index.browser.esm2017.js:17514:19)
    at Observable.eval [as _subscribe] (webpack-internal:///../rxfire/dist/firestore/index.esm.js:43:89)
    at Observable._trySubscribe (webpack-internal:///../rxfire/node_modules/rxjs/dist/esm5/internal/Observable.js:78:25)
    at Observable.subscribe (webpack-internal:///../rxfire/node_modules/rxjs/dist/esm5/internal/Observable.js:43:30)
    at eval (webpack-internal:///../rxfire/node_modules/rxjs/dist/esm5/internal/operators/map.js:12:16)
    at SafeSubscriber.eval (webpack-internal:///../rxfire/node_modules/rxjs/dist/esm5/internal/util/lift.js:16:28)
    at Observable.subscribe (webpack-internal:///../rxfire/node_modules/rxjs/dist/esm5/internal/Observable.js:38:30)
    at eval (webpack-internal:///./pages/index.js:46:275)
    at invokePassiveEffectCreate (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:23482:20)
    at HTMLUnknownElement.callCallback (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:3994:16)
    at invokeGuardedCallback (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:4056:31)
    at flushPassiveEffectsImpl (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:23569:9)
    at unstable_runWithPriority (webpack-internal:///./node_modules/scheduler/cjs/scheduler.development.js:468:12)
    at runWithPriority$1 (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:11276:10)
    at flushPassiveEffects (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:23442:14)
    at eval (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:23319:11)
    at workLoop (webpack-internal:///./node_modules/scheduler/cjs/scheduler.development.js:417:34)
    at flushWork (webpack-internal:///./node_modules/scheduler/cjs/scheduler.development.js:390:14)
    at MessagePort.performWorkUntilDeadline (webpack-internal:///./node_modules/scheduler/cjs/scheduler.development.js:157:27)
```

## Steps to reproduce
1. `npx create-next-app`
2. `npm install rxjs firebase@exp`
3. `git submodule add -b exp https://github.com/FirebaseExtended/rxfire.git rxfire` -> https://github.com/FirebaseExtended/rxfire/tree/da26625dd3448a265262af0d0537c191e20329d0
4. `(cd rxfire && npm install --force && npm run build && npm link)`
5. `npm link rxfire`
6. Replace pages/index.js into this:
    ```javascript
    import React from "react";
    import { getApp, getApps, initializeApp } from "firebase/app";
    import { doc, getFirestore } from "firebase/firestore";
    import { docData } from "rxfire/firestore";

    const firebase = (() => {
      if (getApps().length === 0) {
        return initializeApp({ /* OPTIONS */ });
      } else {
        return getApp();
      }
    })();

    export default function Home() {
      React.useEffect(() => {
        const subscriptions = [];
        subscriptions[subscriptions.length] = docData(doc(getFirestore(firebase), "collections", "id")).subscribe(console.log);
        return () =>
          subscriptions.forEach((subscription) => subscription.unsubscribe());
      }, []);
      return (
        <></>
      )
    }
    ```

File: firestore/fromRef.ts (hunk 1 above)
Comment (jhuleatt) [reply firestore/fromRef.ts]: fixed

**Comment by jhuleatt** [comment]
Thanks for trying it out, @uny! I'm receiving the same error you are (`Uncaught FirebaseError: Expected type '_a', but it was: an object`), but it seems to be build-specific. For example, [this codesandbox](https://codesandbox.io/s/boring-sanne-1j0jt?file=/src/index.ts) works fine. I'll keep looking into it!

**Review by jhuleatt — APPROVED** [review]

**Pull request merged**